firebase-firestore/src/androidTest/java/com/google/firebase/firestore/TransactionTest.java

```diff
@@ -637,6 +705,29 @@ public void testDoesNotRetryOnPermanentError() {
     assertEquals(1, count.get());
   }
 
+  @Test
+  public void testRetryOnAlreadyExistsError() {
+    final FirebaseFirestore firestore = testFirestore();
+    DocumentReference doc = firestore.collection("foo").document();
+    CountDownLatch latch = new CountDownLatch(2);
+    waitFor(
+        firestore.runTransaction(
+            transaction -> {
+              latch.countDown();
+              transaction.get(doc);
+              // Do a write outside of the transaction.
+              waitFor(doc.set(map("count", 1)));
```

Comment:
Shouldn't the test skip writing to the document on the 2nd invocation? Otherwise, wouldn't the 2nd invocation? Otherwise, wouldn't the 2nd transaction commit also fail and get retried?

Reply:
hmm 🤔 , correct me if I am wrong. 
In the first attempt, transaction.get() gets the reference along with "doc doesn't exist"status. We create the doc outside of the transaction, and then try transaction.set(). Here, transaction still thinks "doc doesn't  exist", and it will be rejected the with the "already-exists" error due to the conflict.
In the retry, transaction.get() gets the reference to the doc, which exists this time, we update the doc outside the transaction, and then transaction tries transaction.set(). As there are no conflicts of existence this time, it is going to succeed.

Reply:
In case anyone is following this conversation, the resolution was that the subsequent call to `doc.set(map("count", 1))` was elided by the backend and did not affect the document's `updateTime`. As a result, the test actually passed. If the value `1` was changed to something dynamic, like `retryCount.get()`, then the test would fail.